neps/nep-0488.md

```diff
@@ -0,0 +1,1537 @@
+
+- The input length is not divided by 224
+- Too much memory is used
+- Field elements encoded incorrectly (see Curve points encoded section)
+- Any points not on the curve
+
+***Annotation:***
+
+```rust
+/// Computes multiexp on BLS12-381 curve using Pippenger's algorithm 
+///\sum_i si*pi should be equal result.
+///
+/// # Arguments
+///
+/// * `value` - sequence of (pi, si), where
+///    pi is point (x:Fp^2, y:Fp^2) on BLS12-381, and si is u256.
+///    BLS12-381 is Y^2 = X^3 + 4(i + 1) curve over Fp^2.
+///
+///   `value` is encoded as packed, big-endian
+///   `[(([u8; 96], [u8; 96]), u256)]` slice.
+///
+/// # Errors
+///
+/// If `value_len + value_ptr` points outside the memory or the registers
+/// use more memory than the limit, the function returns
+/// `MemoryAccessViolation`.
+///
+/// If point coordinates are not on curve, point is not in the subgroup,
+/// scalar is not in the field or  `value.len()%224!=0`, the function returns
+/// `BLS12381InvalidInput`.
+///
+/// # Cost
+///
+/// `base + write_register_base + write_register_byte * num_bytes +
+///  bls12381_g2_multiexp_base +
+///  bls12381_g2_multiexp_element * num_elements + 
+///  bls12381_g2_multiexp_element_div_log * num_elements/max(1, log(num_elements))`
+pub fn bls12381_g2_multiexp(
+        &mut self,
+        value_len: u64,
+        value_ptr: u64,
+        register_id: u64,
+) -> Result<()>;
+```
+
+#### bls12381_map_fp_to_g1
+
+***Description:***
+
+The function takes the element $a \in F_p$ and maps it to the $G_1 \subset E(F_p)$. The specification of the mapping function you can find in the section "Map to curve specification". This function does NOT perform mapping of the byte string into $F_p$, it can be implemented in different ways and this can be performed effectively in contract.
+
+***Input:*** the function takes as input `48` bytes — the element from $F_p$ (one unsigned integer $< p$). More details are in the Curve Points Encoding section.
+
+***Output:*** the output is `96 bytes` — one point $\in G_1 \subset E(F_p)$ in decompressed format. More details are in the Curve Points Encoding section.
+
+***Gas Estimation:***
+
+The gas consumption is a constant calculated empirically.
+
+***Test cases:***
+
+- Correct $F_p$ element
+- $a = 0$
+- $a \ge p$
+- Edge cases for inner algorithms for mapping[^49]
+
+***Tests References:*** The EIP-2537 contains the same function, so we can reuse test vectors from Ethereum[^47],[^48].
+
+***Error cases:***
+
+- Incorrect input length
+- $a \ge p$
+
+***Annotation:***
+
+```rust
+/// Map elements from Fp to G1 subset of BLS12-381 curve
+///
+/// # Arguments
+///
+/// * `value` - 48 bytes, the element from Fp -- unsigned integer < p 
+///
+/// # Errors
+///
+/// If `value_len + value_ptr` points outside the memory or the registers
+/// use more memory than the limit, the function returns
+/// `MemoryAccessViolation`.
+///
+/// If value_len != 48 or the value >= p the function return `BLS12381InvalidInput`.
```

Comment:
Why have a parameter, if it’s to hardcode it to a single valid value? Do you foresee extensions where this value would change?

Reply:
Because if the value comes from a register, then `value_len` would be `u64::MAX`, and the code would use the length of the value in the register instead.

Reply:
Wait, is values coming from registers part of the specification proposed? As far as I can tell this function needs a `value_ptr`, which means it must be somewhere in the memory of the wasm program.

Reply:
Generally NEAR host functions obtain input from memory by using `get_memory_or_register!`, which can also get it from a register.

Reply:
This is "generally", if we do want it it needs to be explicitly listed in the NEP. And I agree that we probably do want it considering our other host functions, but the current NEP, as written, must not have an implementation that uses `get_memory_or_register`, as it never mentions being able to take input from registers.

Reply:
@Ekleog-NEAR Hm... The difficult conversations are started. Yes, in implementation I am going to use `get_memory_or_register`. It would be good if you give me an example of how it is explained in another NEP if you have something in your mind. If you don't have a quick answer -- no problem, I will find an example by myself.

Reply:
Unfortunately I think we only recently started to actively use formalized NEP, so I don’t know of a good example. I’d say just a short sentence in the definition of each function that takes address+len, that says that if len is u64::max then registers are used to fetch the data, and the length of the register is used, should I think be enough to be explicit that ptr+len is not only used to refer to the memory address space :)